src/radical/pilot/resource_config.py

```diff
@@ -194,3 +201,540 @@ class ResourceConfig(ru.TypedDict):
+    UNIFORM        = 'uniform'
+    CORES_PER_NODE = 'cores_per_node'
+    GPUS_PER_NODE  = 'gpus_per_node'
+    LFS_PER_NODE   = 'lfs_per_node'
+    MEM_PER_NODE   = 'mem_per_node'
+
+    _schema = {
+        NODES          : [NodeResources],
+        UNIFORM        : bool,
+        CORES_PER_NODE : int,
+        GPUS_PER_NODE  : int,
+        LFS_PER_NODE   : int,
+        MEM_PER_NODE   : int,
+
+    }
+
+    _defaults = {
+        NODES          : list(),
+        UNIFORM        : True,
+        CORES_PER_NODE : None,
+        GPUS_PER_NODE  : None,
+        LFS_PER_NODE   : None,
+        MEM_PER_NODE   : None,
+    }
+
+    def __init__(self, from_dict: dict = None, **kwargs) -> None:
+
+        super().__init__(from_dict=from_dict, **kwargs)
+
+        self.__index          = int()
+        self.__nodes_by_index = dict()
+        self.__nodes_by_name  = dict()
+        self.__last_failed_rr = None
+        self.__last_failed_n  = None
+
+        self.__verified = False
+
+
+    # --------------------------------------------------------------------------
+    #
+    def verify(self) -> None:
+
+        if not self.nodes:
+            return
+
+        self.uniform = True
+        node_0  = self.nodes[0]
+        for node in self.nodes[1:]:
+
+            if node.cores != node_0.cores or \
+               node.gpus  != node_0.gpus  or \
+               node.lfs   != node_0.lfs   or \
+               node.mem   != node_0.mem:
+                self.uniform = False
+                break
+
+        if self.uniform:
+            self.cores_per_node = len(node_0.cores)
+            self.gpus_per_node  = len(node_0.gpus)
+            self.lfs_per_node   = node_0.lfs
+            self.mem_per_node   = node_0.mem
+
+        else:
+            self.cores_per_node = None
+            self.gpus_per_node  = None
+            self.lfs_per_node   = None
+            self.mem_per_node   = None
+
+        self.__nodes_by_name  = {node.name : node for node in self.nodes}
+
+        self.__verified = True
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _assert_rr(self, rr: RankRequirements, n_slots:int) -> None:
+
+        if not self.__verified:
+            self.verify()
+
+        if not self.uniform:
+            raise RuntimeError('verification unsupported for non-uniform nodes')
+
+        if not rr.n_cores:
+            raise ValueError('invalid rank requirements: %s' % rr)
+
+        ranks_per_node = self.cores_per_node / rr.n_cores
+
+        if rr.n_gpus:
+            ranks_per_node = min(ranks_per_node, self.gpus_per_node / rr.n_gpus)
+
+        if rr.lfs:
+            ranks_per_node = min(ranks_per_node, self.lfs_per_node / rr.lfs)
+
+        if rr.mem:
+            ranks_per_node = min(ranks_per_node, self.mem_per_node / rr.mem)
+
+        if ranks_per_node < 1:
+            raise ValueError('invalid rank requirements: %s' % rr)
+
+        if int(ranks_per_node) * n_slots > len(self.nodes):

```

Comment:
something is wrong with this comparison - isn't `ranks_per_node` is a max number of slots per node? thus `math.ceil(n_slots / ranks_per_node)` is the number of required nodes?

Reply:
Ack!  Seems I was rather sloppy with all this... :-P